<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A Tool for Mass Spectrometry–Based Ecological Metabolomics • ecomet</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Tool for Mass Spectrometry–Based Ecological Metabolomics">
<meta name="description" content="The ecomet package provides tools for processing and analyzing mass spectrometry–based metabolomics data in ecological and evolutionary contexts. It establishes a standardized pipeline that integrates MS1 feature–abundance tables with MS2 spectral similarity to generate common data products, including principal component analyses (PCA), chemical dendrograms where tips represent compounds, principal coordinates analyses (PCoA), and differential expression analyses. Unlike existing metabolomics toolkits that focus primarily on biomedical or cheminformatics applications, ecomet is designed explicitly for ecometabolomics. It emphasizes workflows that link metabolomic variation to ecological data, making it easier to move from raw mass spectrometry files to reproducible, comparative analyses of plant chemical diversity.">
<meta property="og:description" content="The ecomet package provides tools for processing and analyzing mass spectrometry–based metabolomics data in ecological and evolutionary contexts. It establishes a standardized pipeline that integrates MS1 feature–abundance tables with MS2 spectral similarity to generate common data products, including principal component analyses (PCA), chemical dendrograms where tips represent compounds, principal coordinates analyses (PCoA), and differential expression analyses. Unlike existing metabolomics toolkits that focus primarily on biomedical or cheminformatics applications, ecomet is designed explicitly for ecometabolomics. It emphasizes workflows that link metabolomic variation to ecological data, making it easier to move from raw mass spectrometry files to reproducible, comparative analyses of plant chemical diversity.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">ecomet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/vignette-eCOMET-intro.html">vignette-eCOMET-intro</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="ecomet">eCOMET<a class="anchor" aria-label="anchor" href="#ecomet"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<p>The eCOMET package provides tools for processing and analyzing mass spectrometry–based metabolomics data in ecological and evolutionary contexts. It establishes a standardized pipeline that integrates MS1 feature–abundance tables with MS2 spectral similarity to generate common data products, including principal component analyses (PCA), chemical dendrograms where tips represent compounds, principal coordinates analyses (PCoA), and differential expression analyses. Unlike existing metabolomics toolkits that focus primarily on biomedical or cheminformatics applications, ecomet is designed explicitly for ecometabolomics. It emphasizes workflows that link metabolomic variation to ecological data, making it easier to move from raw mass spectrometry files to reproducible, comparative analyses of plant chemical diversity.</p>
<p>The major toolboxes of eCOMET are as follows: <img width="1921" height="1080" alt="Screenshot 2025-07-24 at 6 59 47 PM" src="https://github.com/user-attachments/assets/517b33f9-7b66-4067-bda8-a66ae0a1d99c"></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of ecomet from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("pak")</span></span>
<span><span class="fu">pak</span><span class="fu">::</span><span class="fu"><a href="https://pak.r-lib.org/reference/pak.html" class="external-link">pak</a></span><span class="op">(</span><span class="st">"dlforrister/ecoMET_webpage"</span><span class="op">)</span></span></code></pre></div>
<p>This is a basic example which shows you how to solve a common problem:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dlforrister.github.io/ecoMET_webpage/">ecomet</a></span><span class="op">)</span></span>
<span><span class="co">## basic example code</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="id_0-input">0. Input<a class="anchor" aria-label="anchor" href="#id_0-input"></a>
</h2>
<p>Following input files are required. See ‘/raw_data’ of demo files. 1) Feature quantification table generated by MZMine4 (<code>features.csv</code>) This file is generated by [Feature list methods → Export feature list → CSV]. See <a href="https://mzmine.github.io/mzmine_documentation/module_docs/io/feat-list-export.html" class="external-link uri">https://mzmine.github.io/mzmine_documentation/module_docs/io/feat-list-export.html</a> 2) Annotation of features by SIRIUS (<code>canopus_formula_summary.tsv</code> and <code>structure_identification.tsv</code>) These are included in the standard output of SIRIUS. Make sure you run all SIRIUS → ZODIAC → CSI:FingerID → CANOPUS. Then export as tsv. 3) Metadata of samples (<code>metadata.csv</code>) Metadata should have three or more columns; the first three being sample, group, and mass. Additional information can be provided (see Part 9. Regression with metadata). 4) (Optional) Custom annotation of features (‘custom_DB_glucosinolates.csv’) The custom DB includes three columns: compound, mz, rt 5) (Optional) Chemical similarity table from MZMine 4. The metric can be cosine, DreaMS, and MS2DeepScore (cosine.csv, dreams.csv, and ms2deepscore.csv) To obtain these data, you need to run molecular networking at MZMine4 for each metric. See <a href="https://mzmine.github.io/mzmine_documentation/module_docs/group_spectral_net/molecular_networking.html" class="external-link uri">https://mzmine.github.io/mzmine_documentation/module_docs/group_spectral_net/molecular_networking.html</a> Set the minimal similarity value as 0 to get all pairwise similarity. These tables are impoted and processed for heatmap visualization and chemical diversity quantification (see 10. Chemical Diversity Measures)</p>
<p>The demo files are metabolomics analysis files of Arabidopsis thaliana (Col-0) attacked by two different herbivores (Spodoptera litura; sl, and Lipaphis erysimi; le). Eight replicates were sampled and analyzed for each group. See metadata.csv for more details.</p>
</div>
<div class="section level2">
<h2 id="id_1-initializing-the-mmo-object">1. Initializing the MMO object<a class="anchor" aria-label="anchor" href="#id_1-initializing-the-mmo-object"></a>
</h2>
<p>All input files are imported to construct a <code>mmo</code> object. The <code>mmo</code> object will be used in downstream statistics.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="do">###</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co"># 1.1. Give directories</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>mzmine_featuredir <span class="ot">&lt;-</span> <span class="st">'raw_data/250724_features_ms2.csv'</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>metadatadir <span class="ot">&lt;-</span> <span class="st">"raw_data/250728_mmo_metadata.csv"</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>canopus_formuladir <span class="ot">&lt;-</span> <span class="st">"raw_data/canopus_formula_summary.tsv"</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>canopus_structuredir <span class="ot">&lt;-</span> <span class="st">"raw_data/structure_identifications.tsv"</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>gls_db <span class="ot">&lt;-</span> <span class="st">'raw_data/250428_GLS_Ath_simplegrad_MMO.csv'</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>cos_dir <span class="ot">&lt;-</span> <span class="st">'raw_data/250728_sim_ms2_(modified)_cosine.csv'</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>dreams_dir <span class="ot">&lt;-</span> <span class="st">'raw_data/250728_sim_dreams.csv'</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>m2ds_dir <span class="ot">&lt;-</span> <span class="st">'raw_data/250728_sim_ms2deepscore.csv'</span></span></code></pre></div>
<p>First import the mzmine feature list and metadata to create a <code>mmo</code> object.</p>
<p><code>mmo &lt;- GetMZmineFeature(mzmine_featuredir, metadatadir) head(mmo$featre_data)</code></p>
<p>then add annotation generated from sirius to the object</p>
<p><code>mmo &lt;- AddSiriusAnnot(mmo, canopus_structuredir, canopus_formuladir) head(mmo$sirius_annot)</code></p>
<p>Custom annotation can be added based on m/z and RT.</p>
<p><code>mmo &lt;- AddCustomAnnot(mmo, DB = gls_db, mztol = 5, rttol = 0.1) # Add annotation</code></p>
<p>then process the quantification data</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Normalize data</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>mmo <span class="ot">&lt;-</span> <span class="fu">ReplaceZero</span>(mmo, <span class="at">method =</span> <span class="st">'one'</span>) <span class="co"># Replace 0 and NA values by 1</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>mmo <span class="ot">&lt;-</span> <span class="fu">MassNormalization</span>(mmo) <span class="co"># Normalize peak area by sample mass in metadata</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>mmo <span class="ot">&lt;-</span> <span class="fu">MeancenterNormalization</span>(mmo) <span class="co"># Add mean-centered area</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>mmo <span class="ot">&lt;-</span> <span class="fu">LogNormalization</span>(mmo) <span class="co"># Add log-transformed area</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>mmo <span class="ot">&lt;-</span> <span class="fu">ZNormalization</span>(mmo) <span class="co"># Add Zscore</span></span></code></pre></div>
<p>The chemical similarity tables are imported then transformed into distance matrix</p>
<p><code># Import chemical distance data for chemical diversity analyses mmo &lt;- AddChemDist(mmo, cos_dir = cos_dir, dreams_dir = dreams_dir, m2ds_dir = m2ds_dir)</code></p>
<p>Check every data are imported well</p>
<p><code>summary(mmo)</code></p>
<div class="section level3">
<h3 id="mmo-object">mmo object<a class="anchor" aria-label="anchor" href="#mmo-object"></a>
</h3>
<p>The <code>mmo</code> object contains following information 1) Feature quantification value (<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>f</mi><mi>e</mi><mi>a</mi><mi>t</mi><mi>u</mi><mi>r</mi><msub><mi>e</mi><mi>d</mi></msub><mi>a</mi><mi>t</mi><mi>a</mi><mo stretchy="false" form="postfix">)</mo><mn>2</mn><mo stretchy="false" form="postfix">)</mo><mi>M</mi><mi>e</mi><mi>t</mi><mi>a</mi><mi>d</mi><mi>a</mi><mi>t</mi><mi>a</mi><mo stretchy="false" form="prefix">(</mo></mrow><annotation encoding="application/x-tex">feature_data)
2) Metadata (</annotation></semantics></math>metadata) 3) Pairwise comparison data (<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>p</mi><mi>a</mi><mi>i</mi><mi>r</mi><mi>w</mi><mi>i</mi><mi>s</mi><mi>e</mi><mo stretchy="false" form="postfix">)</mo><mn>4</mn><mo stretchy="false" form="postfix">)</mo><mi>L</mi><mi>o</mi><mi>g</mi><mo>,</mo><mi>m</mi><mi>e</mi><mi>a</mi><mi>n</mi><mo>−</mo><mi>c</mi><mi>e</mi><mi>n</mi><mi>t</mi><mi>e</mi><mi>r</mi><mi>e</mi><mi>d</mi><mo>,</mo><mi>a</mi><mi>n</mi><mi>d</mi><mi>Z</mi><mo>−</mo><mi>t</mi><mi>r</mi><mi>a</mi><mi>n</mi><mi>s</mi><mi>f</mi><mi>o</mi><mi>r</mi><mi>m</mi><mi>e</mi><mi>d</mi><mi>f</mi><mi>e</mi><mi>a</mi><mi>t</mi><mi>u</mi><mi>r</mi><mi>e</mi><mi>q</mi><mi>u</mi><mi>a</mi><mi>n</mi><mi>t</mi><mi>i</mi><mi>f</mi><mi>i</mi><mi>c</mi><mi>a</mi><mi>t</mi><mi>i</mi><mi>o</mi><mi>n</mi><mi>v</mi><mi>a</mi><mi>l</mi><mi>u</mi><mi>e</mi><mo stretchy="false" form="prefix">(</mo></mrow><annotation encoding="application/x-tex">pairwise)
4) Log, mean-centered, and Z-transformed feature quantification value (</annotation></semantics></math>log, <span class="math inline">$meancentered, and #zscore)
5) Sirius annotation ($</span>sirius_annot) 6) cosine distance (<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mi>o</mi><mi>s</mi><mi>.</mi><mi>d</mi><mi>i</mi><mi>s</mi><mi>s</mi><mi>i</mi><mi>m</mi><mo stretchy="false" form="postfix">)</mo><mn>7</mn><mo stretchy="false" form="postfix">)</mo><mi>D</mi><mi>r</mi><mi>e</mi><mi>a</mi><mi>M</mi><mi>S</mi><mi>d</mi><mi>i</mi><mi>s</mi><mi>t</mi><mi>a</mi><mi>n</mi><mi>c</mi><mi>e</mi><mo stretchy="false" form="prefix">(</mo></mrow><annotation encoding="application/x-tex">cos.dissim)
7) DreaMS distance (</annotation></semantics></math>dreams.dissim) 8) MS2DeepScore distance (m2ds.dissim)</p>
<p>All features have id (from mzmine) and feature (mz_rt generated by GetMZmineFeature()). Both can be interchangable by FeatureToID(mmo, features) and IDToFeature(mmo, IDs).</p>
</div>
<div class="section level3">
<h3 id="annotated-features-by-sirius">Annotated features by SIRIUS<a class="anchor" aria-label="anchor" href="#annotated-features-by-sirius"></a>
</h3>
<p>Some feature groups may deserve particular interest. For the demo file, glucosinolates and flavonoids are. We can briefly annotate those metabolites using annotation from sirius.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>features <span class="ot">&lt;-</span> mmo<span class="sc">$</span>sirius_annot</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co"># Get list of glucosinolated using custom annoatation DB</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>gls_hits <span class="ot">&lt;-</span> mmo<span class="sc">$</span>custom_annot <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">lengths</span>(custom_annot) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>GLSs <span class="ot">&lt;-</span> gls_hits <span class="sc">%&gt;%</span> <span class="fu">pull</span>(feature)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># Get list of flavonoids using sirius annotation</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>FLVs <span class="ot">&lt;-</span> features <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">str_detect</span>(features[[<span class="dv">46</span>]], <span class="st">"Flavonoid"</span>)) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(feature)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="id_2-summarise-using-pls-da">2. Summarise using PLS-DA<a class="anchor" aria-label="anchor" href="#id_2-summarise-using-pls-da"></a>
</h2>
<p>By using PLS-DA plot, the general distribution of metabolic profile of each sample and group can be visualized. For each group give colors</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Define your custom colors for each group</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>custom_colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ctrl"</span> <span class="ot">=</span> <span class="st">"#999999"</span>, <span class="st">"sl1"</span> <span class="ot">=</span> <span class="st">"#fdcdac"</span>, <span class="st">"le1"</span> <span class="ot">=</span> <span class="st">"#b3e2cd"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#Or automatically give colors</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>custom_colors <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">brewer.pal</span>(<span class="fu">length</span>(<span class="fu">unique</span>(mmo<span class="sc">$</span>metadata<span class="sc">$</span>group)), <span class="st">"Set3"</span>), <span class="fu">unique</span>(mmo<span class="sc">$</span>metadata<span class="sc">$</span>group))</span></code></pre></div>
<p>Then plot PLS-DA</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">PLSDAplot</span>(mmo, <span class="at">color =</span> custom_colors, <span class="at">outdir =</span> <span class="st">'plots/plsda/PLSDA_Z.pdf'</span>, <span class="at">normalization =</span> <span class="st">'Z'</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">PLSDAplot</span>(mmo, <span class="at">color =</span> custom_colors, <span class="at">topk =</span> <span class="dv">0</span>, <span class="at">outdir =</span> <span class="st">'plots/plsda/PLSDA_Z_noLoadings.pdf'</span>, <span class="at">normalization =</span> <span class="st">'Z'</span>)</span></code></pre></div>
<p><img width="603" height="608" alt="Screenshot 2025-09-01 at 9 15 29 AM" src="https://github.com/user-attachments/assets/01630a20-c7b3-46d8-833e-85291ab9df37"></p>
<p>The <code>topk</code> parameter can be adjusted to plot loadings of top features on the plot. The <code>normalization</code> can be <code>None</code>, <code>Log</code>, <code>Meancentered</code>, and <code>Z</code>. If a subset of features are to be used for ploting, set <code>filter_feature = TRUE</code> then provide the list of features to the <code>feature_list</code> If a subset of groups are to be plotted, set ‘filter_group = TRUE’ then provide the list of groups to the group_list.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Glucosinolates only</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">PLSDAplot</span>(mmo, custom_colors, <span class="at">topk =</span> <span class="dv">0</span>, <span class="at">outdir =</span> <span class="st">'plots/plsda_GLS.pdf'</span>, <span class="at">normalization =</span> <span class="st">'Z'</span>, <span class="at">filter_feature =</span> <span class="cn">TRUE</span>, <span class="at">feature_list =</span> GLSs)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="id_3-pairwise-comparison-dams">3. Pairwise Comparison (DAMs)<a class="anchor" aria-label="anchor" href="#id_3-pairwise-comparison-dams"></a>
</h2>
<p>Many analyses targets to find <strong>Differentially Accumulated Metabolites (DAMs)</strong>. DAMs can be defined by thresholds of log2-fold change and adjusted p-value. Those two metrics are calculated by following code. Note that the divisor group is at the left.</p>
<p><code># 3.1. Add pairwise comparisons mmo &lt;- PairwiseComp(mmo, 'ctrl', 'sl1') mmo &lt;- PairwiseComp(mmo, 'ctrl', 'le1')</code></p>
<p>Then DAMs of each comparison can be extracted.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># 3.2. Get DAMs from the comparisons</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>DAMs <span class="ot">&lt;-</span> <span class="fu">GetDAMs</span>(mmo, <span class="at">fc_cutoff =</span> <span class="fl">0.5849625</span>, <span class="at">pval_cutoff =</span> <span class="fl">0.1</span>) <span class="co"># log2(1.5) = 0.5849625</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>DAMs_up <span class="ot">&lt;-</span> DAMs<span class="sc">$</span>DAMs_up</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>DAMs_down <span class="ot">&lt;-</span> DAMs<span class="sc">$</span>DAMs_down</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="fu">head</span>(DAMs_up)</span></code></pre></div>
<p>The cutoffs for DAMs can be adjusted in the GetDAMs function</p>
</div>
<div class="section level2">
<h2 id="id_4-upset-and-venn-diagram">4. Upset and Venn Diagram<a class="anchor" aria-label="anchor" href="#id_4-upset-and-venn-diagram"></a>
</h2>
<p>How many features are overlapping and specific are the common questions following DAM analysis. It can be visualized by UpSet plot and Venn diagram. The inputs for both are identical and can be prepared as following</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># 4.1. Define groups</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>VennInput <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  <span class="at">sl1.up =</span> DAMs_up<span class="sc">$</span>ctrl_vs_sl1.up,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  <span class="at">le1.up =</span> DAMs_up<span class="sc">$</span>ctrl_vs_le1.up</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>)</span></code></pre></div>
<p>Then can be plotted as following</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># 4.2. Plot</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co"># 4.2.1. Venn Diagram</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="fu">ggvenn</span>(VennInput, <span class="at">stroke_size =</span> <span class="fl">0.5</span>, <span class="at">set_name_size =</span> <span class="dv">4</span>, <span class="at">show_percentage =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) </span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"plots/Venn/Venn_Upreg.pdf"</span>, <span class="at">height =</span> <span class="dv">5</span>, <span class="at">width =</span> <span class="dv">5</span>)</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co"># 4.2.2. Upset plot</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"plots/Venn/Upset_Upreg.pdf"</span>, <span class="dv">7</span>, <span class="dv">5</span>)</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(VennInput), <span class="at">nsets=</span><span class="dv">10</span>, <span class="at">nintersects=</span><span class="dv">20</span>,<span class="at">order.by=</span><span class="st">'freq'</span>, <span class="at">mainbar.y.label=</span><span class="st">'Features in Set'</span>, <span class="at">line.size=</span><span class="dv">1</span>, <span class="at">point.size=</span><span class="dv">4</span>, <span class="at">shade.color=</span><span class="st">'white'</span>, <span class="at">text.scale=</span><span class="dv">1</span>, <span class="at">show.numbers=</span><span class="cn">FALSE</span>)</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img width="601" height="419" alt="Screenshot 2025-09-01 at 9 16 00 AM" src="https://github.com/user-attachments/assets/8c92244c-c6c7-4197-983b-03dead991750"><img width="715" height="606" alt="Screenshot 2025-09-01 at 9 16 14 AM" src="https://github.com/user-attachments/assets/a2c8018f-9279-4497-aa06-bc2391638afe"></p>
</div>
<div class="section level2">
<h2 id="id_5-volcano-plot">5. Volcano plot<a class="anchor" aria-label="anchor" href="#id_5-volcano-plot"></a>
</h2>
<p>The pairwise comparion can be visualized by a volcano plot. Following code generates volcano plots for all comparison made in section 3.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>comparison_columns <span class="ot">&lt;-</span> <span class="fu">colnames</span>(mmo<span class="sc">$</span>pairwise)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>log2FC_columns <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"log2FC"</span>, comparison_columns, <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>comparisons <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">"_$"</span>, <span class="st">""</span>, <span class="fu">unique</span>(<span class="fu">sub</span>(<span class="st">"log2FC"</span>, <span class="st">""</span>, log2FC_columns)))  <span class="co"># Remove trailing underscore from comparisons</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="cf">for</span> (comp <span class="cf">in</span> comparisons){</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>  <span class="fu">VolcanoPlot</span>(mmo, <span class="at">comp =</span> comp, <span class="at">topk =</span> <span class="dv">10</span>, <span class="at">outdir =</span> <span class="fu">paste</span>(<span class="st">'plots/volcano/volcano_'</span>, comp, <span class="st">'.pdf'</span>, <span class="at">sep =</span> <span class="st">''</span>))</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>}</span></code></pre></div>
<p><img width="505" height="502" alt="Screenshot 2025-09-01 at 9 16 30 AM" src="https://github.com/user-attachments/assets/082f866b-d8b1-4e4c-b1c3-5d07155ce733"></p>
</div>
<div class="section level2">
<h2 id="id_6-heat-map">6. Heat Map<a class="anchor" aria-label="anchor" href="#id_6-heat-map"></a>
</h2>
<p>A heatmap is a great way to visualize the whole metabolome. Typically the features are clustered by the distribution pattern but alternatively the chemical similarity between features can be visualized. For heatmap, required inputs can be generated from mmo as follows:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># 6.1. Generate inputs for heat map</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>HMinput <span class="ot">&lt;-</span> <span class="fu">GenerateHeatmapInputs</span>(mmo, <span class="at">filter_feature =</span> <span class="cn">FALSE</span>, <span class="at">feature_list =</span> feature_list, </span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>                                <span class="at">filter_group =</span> <span class="cn">FALSE</span>, <span class="at">group_list =</span> group_list, </span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>                                <span class="at">summarize =</span> <span class="st">'mean'</span>, <span class="at">control_group =</span> <span class="st">'ctrl'</span>, </span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>                                <span class="at">normalization =</span> <span class="st">'Z'</span>, <span class="at">distance =</span> <span class="st">'dreams'</span>)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="fu">summary</span>(HMinput)</span></code></pre></div>
<p>Where filter_feature and filter_group can be used on desire. The ‘summarize’ argument can be ‘mean’ or ‘fold_change’. If ‘fold_change’ is used, the ‘control_group’ should be noted. If the chemical similarities are to be plotted, the distance metric to be used should be designated in ‘distance’ as one of dreams, cosine, or m2ds. Then the class-level annotation table is generated.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># 6.2. Generate NPC-based annotation table for heatmap</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>sirius_annot <span class="ot">&lt;-</span> mmo<span class="sc">$</span>sirius_annot</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="co"># Get NPC Annotations</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>NPC_pathway <span class="ot">&lt;-</span> <span class="fu">unique</span>(sirius_annot[[<span class="dv">32</span>]])</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>NPC_pathway <span class="ot">&lt;-</span> NPC_pathway[<span class="sc">!</span><span class="fu">is.na</span>(NPC_pathway) <span class="sc">&amp;</span> NPC_pathway <span class="sc">!=</span> <span class="st">""</span>] <span class="co"># remove NA and empty </span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>NPC_superclass <span class="ot">&lt;-</span> <span class="fu">unique</span>(sirius_annot[[<span class="dv">34</span>]])</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>NPC_superclass <span class="ot">&lt;-</span> NPC_superclass[<span class="sc">!</span><span class="fu">is.na</span>(NPC_superclass) <span class="sc">&amp;</span> NPC_superclass <span class="sc">!=</span> <span class="st">""</span>] <span class="co"># remove NA and empty</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>NPC_class <span class="ot">&lt;-</span> <span class="fu">unique</span>(sirius_annot[[<span class="dv">36</span>]])</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>NPC_class <span class="ot">&lt;-</span> NPC_class[<span class="sc">!</span><span class="fu">is.na</span>(NPC_class) <span class="sc">&amp;</span> NPC_class <span class="sc">!=</span> <span class="st">""</span>] <span class="co"># remove NA and empty</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a>sirius_annot_filtered <span class="ot">&lt;-</span> sirius_annot <span class="sc">%&gt;%</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a>  <span class="co"># select(id = 1, NPC_pathway = 30, NPC_class = 34) %&gt;%</span></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a>  <span class="fu">select</span>(<span class="at">id =</span> <span class="dv">1</span>, <span class="at">NPC_class =</span> <span class="dv">36</span>, <span class="at">NPC_superclass =</span> <span class="dv">34</span>, <span class="at">NPC_pathway =</span> <span class="dv">32</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a>  <span class="co"># select(id = 1, NPC_pathway = 30, NPC_superclass = 32) %&gt;%</span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">%in%</span> <span class="fu">rownames</span>(distance_matrix)) <span class="co"># get features with fingerprints</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="fu">rownames</span>(sirius_annot_filtered) <span class="ot">&lt;-</span> sirius_annot_filtered<span class="sc">$</span>id</span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a>sirius_annot_filtered<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb14-19"><a href="#cb14-19" tabindex="-1"></a></span>
<span id="cb14-20"><a href="#cb14-20" tabindex="-1"></a>ann_colors <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb14-21"><a href="#cb14-21" tabindex="-1"></a>    <span class="at">NPC_pathway =</span> <span class="fu">setNames</span>(<span class="fu">brewer.pal</span>(<span class="fu">length</span>(NPC_pathway), <span class="st">"Set2"</span>), NPC_pathway),</span>
<span id="cb14-22"><a href="#cb14-22" tabindex="-1"></a>    <span class="at">NPC_class =</span> <span class="fu">setNames</span>(<span class="fu">viridis</span>(<span class="fu">length</span>(NPC_class)), NPC_class),</span>
<span id="cb14-23"><a href="#cb14-23" tabindex="-1"></a>    <span class="at">NPC_superclass =</span> <span class="fu">setNames</span>(<span class="fu">viridis</span>(<span class="fu">length</span>(NPC_superclass)), NPC_superclass)</span>
<span id="cb14-24"><a href="#cb14-24" tabindex="-1"></a>    )</span></code></pre></div>
<p>The ann_colors will be plotted along with the heapmap to show which chemical class each features are included in. The heatmap can be visualized as follows:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">"plots/heatmap/dreams_total_Z.pdf"</span>, <span class="at">width =</span> <span class="dv">15</span>, <span class="at">height =</span> <span class="dv">20</span>)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="at">mat =</span> HMinput<span class="sc">$</span>FC_matrix, </span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>     <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>, <span class="co">#do not change</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>     <span class="at">clustering_distance_rows =</span> HMinput<span class="sc">$</span>dist_matrix, <span class="co"># Delete this line for UPGMA clustering of rows</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>     <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>, </span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>     <span class="at">clustering_method =</span> <span class="st">"average"</span>, <span class="co">#UPGMA</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>     <span class="at">show_rownames =</span> <span class="cn">TRUE</span>, </span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a>     <span class="at">show_colnames =</span> <span class="cn">TRUE</span>,</span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a>     <span class="at">annotation_row =</span> sirius_annot_filtered, <span class="co"># Dataframe with the rownames are identical with 'mat' and gives annotation</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a>     <span class="at">annotation_colors =</span> ann_colors,</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a>     <span class="at">cellwidth =</span> <span class="dv">25</span>,</span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a>     <span class="at">cellheight =</span> <span class="fl">0.05</span>,</span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a>     <span class="at">treeheight_row =</span> <span class="dv">100</span>,</span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a>     <span class="at">fontsize_row =</span> <span class="dv">3</span>,</span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a>     <span class="at">fontsize_col =</span> <span class="dv">15</span>,</span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a>     <span class="at">scale =</span> <span class="st">'none'</span>,</span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a>     <span class="at">annotation_names_row =</span> <span class="cn">TRUE</span>,</span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a>    <span class="co">#  labels_row = row_label,</span></span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a>     <span class="at">border_color =</span> <span class="cn">NA</span>,</span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a>     <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span>))(<span class="dv">100</span>))</span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="id_7-chemical-class-enrichment-analysis">7. Chemical class enrichment analysis<a class="anchor" aria-label="anchor" href="#id_7-chemical-class-enrichment-analysis"></a>
</h2>
<p>Biological questions ask which class of chemical compounds are enriched in a set of compounds of interest (e.g., DAMs from above). This is analogue to the Gene Ontology enrichment analysis performed in transcriptomics. In MMO, NPC and Classyfire terms annotated by Canopus of SIRIUS are used to perform chemical class enrichment analysis of given list of features. The enrichment score of each term is calculated to plot the number of each term and the significance.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># 7.1. For a single set of features, a detailed enrichment plot can be generated</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="co"># There are two plotting styles available</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="fu">CanopusListEnrichmentPlot</span>(mmo, DAMs_up<span class="sc">$</span>ctrl_vs_sl1.up, <span class="at">pthr =</span> <span class="fl">0.1</span>, <span class="at">outdir =</span> <span class="st">'plots/enrichment/sl1_up.pdf'</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="fu">CanopusListEnrichmentPlot_2</span>(mmo, DAMs_up<span class="sc">$</span>ctrl_vs_sl1.up, <span class="at">pthr =</span> <span class="fl">0.1</span>, <span class="at">outdir =</span> <span class="st">'plots/enrichment/sl1_up_2.pdf'</span>, <span class="at">topn =</span> <span class="dv">10</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img width="608" height="608" alt="Screenshot 2025-09-01 at 9 17 04 AM" src="https://github.com/user-attachments/assets/66bb1636-6baa-4c08-b70a-70a7dde276bf"></p>
<div class="sourceCode" id="cb17"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># 7.2. For a list of sets features, a summary enrichment plot can be generated</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="co"># The summary enrichment plot can be generated for either a single level of CANOPUS classification (8.2.1) or for all levels (8.2.2)</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="co"># 7.2.1. For a single level of CANOPUS classification</span></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>term_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'NPC_class'</span>, <span class="st">'NPC_superclass'</span>, <span class="st">'NPC_pathway'</span>, <span class="st">'ClassyFire_most_specific'</span>, <span class="st">'ClassyFire_level5'</span>, <span class="st">'ClassyFire_subclass'</span>, <span class="st">'ClassyFire_class'</span>, <span class="st">'ClassyFire_superclass'</span>)</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="fu">CanopusLevelEnrichmentPlot</span>(mmo, DAMs_up, <span class="at">term_level =</span> <span class="st">'NPC_class'</span>, <span class="at">pthr =</span> <span class="fl">0.1</span>, <span class="at">prefix =</span> <span class="st">'plots/enrichment/DAMs_up_NPC_class.pdf'</span>)</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a><span class="co"># 7.2.2. For all levels of CANOPUS classification</span></span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a><span class="co"># All levels, or onlt NPC or ClassyFire</span></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a>terms <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'NPC'</span>, <span class="st">'ClassyFire'</span>, <span class="st">'all_terms'</span>)</span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="fu">CanopusAllLevelEnrichmentPlot</span>(mmo, DAMs_up, <span class="at">term_level =</span> <span class="st">'all_terms'</span>, <span class="at">pthr =</span> <span class="fl">0.1</span>, <span class="at">prefix =</span> <span class="st">'plots/enrichment/DAMs_up_all_terms'</span>, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">12</span>)</span></code></pre></div>
<p><img width="502" height="501" alt="Screenshot 2025-09-01 at 9 17 23 AM" src="https://github.com/user-attachments/assets/0dac0d0d-8078-496d-8e4d-8285f74e475e"></p>
</div>
<div class="section level2">
<h2 id="id_8-regression-with-metadata">8. Regression with Metadata<a class="anchor" aria-label="anchor" href="#id_8-regression-with-metadata"></a>
</h2>
<p>We are interested in finding anti-herbivore resistive compounds from the plant metabolome and testing whether such compounds are upregulated by insect attack. To do so, we first fit linear mixed model to the amount of each feature and the herbivore performance fed on each plant sample (see metadata). The negative effect size of the model represents the resistive value of the feature. We then test whether resistive features are upregulated (log2FC &gt; 1) by plotting the effect size of the LMM and the log2FC as scatter plot. In the demo file, resistive compounds (which have neative effect sizes) were upregulated, which implies plants produce resistive compounds properly in response to sl attack. The analysis can be done as follows:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># To perform regression, the phenotype of interest should be defined in the metadata as a column</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="co"># 8.1. Regression of individual feature against a phenotype</span></span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="co"># model can be 'lm' or 'pearson' or 'lmm'</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="co"># groups can be a vector of group names, or a single group name which has the phenotpye values in the metadata</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="fu">FeaturePerformanceRegression</span>(mmo, <span class="at">target =</span> <span class="st">'477.0636_7.5687'</span>, <span class="at">phenotype =</span> <span class="st">'sl'</span>, <span class="at">groups =</span> <span class="fu">c</span>(<span class="st">'sl1'</span>), </span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">'lm'</span>, <span class="at">normalization =</span> <span class="st">'Z'</span>, </span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a>  <span class="at">output =</span> <span class="fu">paste</span>(<span class="st">'plots/phenotype_regression/477.0636_7.5687_sl_lm.pdf'</span>))</span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="co"># 8.2. Regression of all features against a phenotype and plotting the results </span></span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a>sl.lm <span class="ot">&lt;-</span> <span class="fu">GetPerformanceFeatureRegression</span>(mmo, <span class="at">phenotype =</span> <span class="st">'sl'</span>, <span class="at">groups =</span> <span class="fu">c</span>(<span class="st">'sl1'</span>), </span>
<span id="cb18-13"><a href="#cb18-13" tabindex="-1"></a>                                        <span class="at">DAM.list =</span> <span class="fu">list</span>(<span class="at">sl.up =</span> DAMs_up<span class="sc">$</span>ctrl_vs_sl1.up, <span class="at">sl.down =</span> DAMs_down<span class="sc">$</span>ctrl_vs_sl1.down), <span class="at">comparisons =</span> <span class="fu">c</span>(<span class="st">'ctrl_vs_sl1'</span>))</span>
<span id="cb18-14"><a href="#cb18-14" tabindex="-1"></a>sl.lm.sig <span class="ot">&lt;-</span> sl.lm <span class="sc">%&gt;%</span> <span class="fu">filter</span>(p_value <span class="sc">&lt;</span> <span class="fl">0.1</span>)</span>
<span id="cb18-15"><a href="#cb18-15" tabindex="-1"></a><span class="fu">head</span>(sl.lm.sig)</span>
<span id="cb18-16"><a href="#cb18-16" tabindex="-1"></a></span>
<span id="cb18-17"><a href="#cb18-17" tabindex="-1"></a>sl.cor <span class="ot">&lt;-</span> <span class="fu">GetPerformanceFeatureCorrelation</span>(mmo, <span class="at">phenotype =</span> <span class="st">'sl'</span>, <span class="at">groups =</span> <span class="fu">c</span>(<span class="st">'sl1'</span>), </span>
<span id="cb18-18"><a href="#cb18-18" tabindex="-1"></a>                                        <span class="at">DAM.list =</span> <span class="fu">list</span>(<span class="at">sl.up =</span> DAMs_up<span class="sc">$</span>ctrl_vs_sl1.up, <span class="at">sl.down =</span> DAMs_down<span class="sc">$</span>ctrl_vs_sl1.down), <span class="at">comparisons =</span> <span class="fu">c</span>(<span class="st">'ctrl_vs_sl1'</span>))</span>
<span id="cb18-19"><a href="#cb18-19" tabindex="-1"></a>sl.cor.sig <span class="ot">&lt;-</span> sl.cor <span class="sc">%&gt;%</span> <span class="fu">filter</span>(p_value <span class="sc">&lt;</span> <span class="fl">0.1</span>)</span>
<span id="cb18-20"><a href="#cb18-20" tabindex="-1"></a><span class="fu">head</span>(sl.cor.sig)</span>
<span id="cb18-21"><a href="#cb18-21" tabindex="-1"></a></span>
<span id="cb18-22"><a href="#cb18-22" tabindex="-1"></a><span class="co"># One may want to plot the regression results along with the fold change values</span></span>
<span id="cb18-23"><a href="#cb18-23" tabindex="-1"></a><span class="fu">PlotFoldchangeResistanceRegression</span>(sl.lm.sig, <span class="at">fold_change =</span> <span class="st">'ctrl_vs_sl1_log2FC'</span>, </span>
<span id="cb18-24"><a href="#cb18-24" tabindex="-1"></a>                                   <span class="at">color =</span> <span class="fu">c</span>(<span class="st">'sl.up'</span> <span class="ot">=</span> <span class="st">'#d42525ff'</span>, <span class="st">'sl.down'</span> <span class="ot">=</span> <span class="st">'#281e99ff'</span>), </span>
<span id="cb18-25"><a href="#cb18-25" tabindex="-1"></a>                                   <span class="at">output_dir =</span> <span class="st">'plots/phenotype_regression/sl_lm_sig.pdf'</span>)</span></code></pre></div>
<p><img width="608" height="607" alt="Screenshot 2025-09-01 at 9 17 57 AM" src="https://github.com/user-attachments/assets/252801ee-4201-46b4-9cf2-2eb2c3c58b80"></p>
</div>
<div class="section level2">
<h2 id="id_9-chemical-diversity-measures">9. Chemical Diversity Measures<a class="anchor" aria-label="anchor" href="#id_9-chemical-diversity-measures"></a>
</h2>
<p>The chemical diversity is one of the key parameters in ecological studies. MMO quantifies the chemical diversity idices using idea from <code>ChemoDiv</code> package. The fingerprint distances are treated as phylogenetic diversity as in measuring taxonomic diversity. Alpha diversity can be calculated as following.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># 9.1. Alpha diversity by Hill numbers</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="co"># q : Hill number order, 0 for richness, 1 for Shannon, 2 for Simpson</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="co"># mode : weighted for the chemical structure (use distance argument), unweighted for no chemical weight</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="co"># filter_feature : if TRUE, only features in the feature_list are used</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="co"># feature_list : a list of features to filter the alpha diversity calculation</span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>alphadiv <span class="ot">&lt;-</span> <span class="fu">GetAlphaDiversity</span>(mmo, <span class="at">q =</span> <span class="dv">3</span>, <span class="at">normalization =</span> <span class="st">'Log'</span>, <span class="at">mode =</span> <span class="st">'weighted'</span>, <span class="at">distance =</span> <span class="st">'dreams'</span>, <span class="at">filter_feature =</span> <span class="cn">FALSE</span>, <span class="at">feature_list =</span> <span class="cn">NULL</span>)</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="co"># 9.1.1 Plot the alpha diversity</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="fu">ggplot</span>(alphadiv, <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> hill_number)) <span class="sc">+</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a>  <span class="fu">geom_beeswarm</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">'Alpha Diversity by Hill Numbers'</span>, <span class="at">x =</span> <span class="st">'Group'</span>, <span class="at">y =</span> <span class="st">'Hill Number'</span>) <span class="sc">+</span></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>, <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">'plots/alphadiv/dreams_q3_log.pdf'</span>, <span class="at">width =</span> <span class="dv">5</span>, <span class="at">height =</span> <span class="dv">5</span>)</span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a><span class="co"># Test for significant differences between groups with ANOVA</span></span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a>anova <span class="ot">&lt;-</span> <span class="fu">anova_tukey_dunnett</span>(alphadiv, <span class="st">'hill_number ~ group'</span>)</span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a><span class="fu">write_anova</span>(anova, <span class="st">'plots/alphadiv/anova_hill_number.csv'</span>)</span></code></pre></div>
<p><img width="506" height="508" alt="Screenshot 2025-09-01 at 9 18 39 AM" src="https://github.com/user-attachments/assets/4f053178-0888-4ef0-b0cb-8ce7460d64d7"></p>
<div class="sourceCode" id="cb20"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># 9.1.2 Plot the feature distribution</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="co"># Function to plot histogram of values for each feature for given groups</span></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>groups <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'ctrl'</span>, <span class="st">'sl1'</span>, <span class="st">'le1'</span>)</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>group.mean <span class="ot">&lt;-</span> <span class="fu">GetGroupMeans</span>(mmo, <span class="at">normalization =</span> <span class="st">'Z'</span>, <span class="at">filter_feature =</span> <span class="cn">FALSE</span>, <span class="at">feature_list =</span> <span class="cn">NULL</span>)[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a>long.group.mean <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">value =</span> <span class="fu">double</span>(), <span class="at">group =</span> <span class="fu">character</span>())</span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="cf">for</span> (group <span class="cf">in</span> groups) {</span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>  group_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">value =</span> group.mean[, group], <span class="at">group =</span> group)</span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a>  <span class="fu">colnames</span>(group_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'value'</span>, <span class="st">'group'</span>)</span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a>  long.group.mean <span class="ot">&lt;-</span> <span class="fu">rbind</span>(long.group.mean, group_data)</span>
<span id="cb20-10"><a href="#cb20-10" tabindex="-1"></a>}</span>
<span id="cb20-11"><a href="#cb20-11" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> long.group.mean, <span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">fill =</span> group, <span class="at">color =</span> group)) <span class="sc">+</span></span>
<span id="cb20-13"><a href="#cb20-13" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">position =</span> <span class="st">'identity'</span>, <span class="at">alpha =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb20-14"><a href="#cb20-14" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb20-15"><a href="#cb20-15" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Normalized peak intensity"</span>, <span class="at">y =</span> <span class="st">"Density"</span>) <span class="sc">+</span></span>
<span id="cb20-16"><a href="#cb20-16" tabindex="-1"></a>  <span class="co"># scale_fill_manual(values = custom_colors) +</span></span>
<span id="cb20-17"><a href="#cb20-17" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span>
<span id="cb20-18"><a href="#cb20-18" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">'plots/alphadiv/density_function_Z.pdf'</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img width="613" height="609" alt="Screenshot 2025-09-01 at 9 18 57 AM" src="https://github.com/user-attachments/assets/83419d41-76e5-49ea-875d-cf2f5148a67f"></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> mmo<span class="sc">$</span>metadata</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>feature <span class="ot">&lt;-</span> mmo<span class="sc">$</span>zscore</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="co"># The distribution should be shown!</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">value =</span> <span class="fu">double</span>(), <span class="at">rank =</span> <span class="fu">double</span>(), <span class="at">sample =</span> <span class="fu">character</span>(), <span class="at">group =</span> <span class="fu">character</span>())</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="cf">for</span> (group <span class="cf">in</span> groups) {</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>  group_samples <span class="ot">&lt;-</span> metadata <span class="sc">%&gt;%</span> <span class="fu">filter</span>(group <span class="sc">==</span> <span class="sc">!!</span>group) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(sample)</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>  <span class="cf">for</span> (sample <span class="cf">in</span> group_samples) {</span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a>    sample_data <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">as.vector</span>(feature[, sample]))</span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a>    sorted_data <span class="ot">&lt;-</span> <span class="fu">sort</span>(sample_data, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a>    sample_plot_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">value =</span> sorted_data, <span class="at">rank =</span> <span class="fu">seq_along</span>(sorted_data), <span class="at">group =</span> group, <span class="at">sample =</span> sample)</span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a>    plot_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(plot_data, sample_plot_data)</span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a>  }</span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a>}</span>
<span id="cb21-14"><a href="#cb21-14" tabindex="-1"></a></span>
<span id="cb21-15"><a href="#cb21-15" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" tabindex="-1"></a></span>
<span id="cb21-17"><a href="#cb21-17" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data, <span class="fu">aes</span>(<span class="at">x =</span> rank, <span class="at">y =</span> value, <span class="at">color =</span> group)) <span class="sc">+</span></span>
<span id="cb21-18"><a href="#cb21-18" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun.data =</span> mean_se, <span class="at">geom =</span> <span class="st">"errorbar"</span>, <span class="at">width =</span> <span class="fl">0.2</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.5</span>), <span class="at">alpha =</span> <span class="fl">0.06</span>) <span class="sc">+</span></span>
<span id="cb21-19"><a href="#cb21-19" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> mean, <span class="at">geom =</span> <span class="st">"line"</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb21-20"><a href="#cb21-20" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb21-21"><a href="#cb21-21" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Sorted Feature"</span>, <span class="at">x =</span> <span class="st">"Rank"</span>, <span class="at">y =</span> <span class="st">"Value"</span>)</span>
<span id="cb21-22"><a href="#cb21-22" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">'plots/alphadiv/sorted_intensity_Z.png'</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img width="960" height="955" alt="Screenshot 2025-09-01 at 9 19 19 AM" src="https://github.com/user-attachments/assets/1c6126f6-f5c9-41b2-ba3b-a05ec6aeef01"></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># 9.2. Beta diversity</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="co"># Calculate beta diversity for different normalizations and methods</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="co"># For unweighted beta diversity, Bray-Curtis or Jaccard distance can be used</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>bray <span class="ot">&lt;-</span> <span class="fu">GetBetaDiversity</span>(mmo, <span class="at">method =</span> <span class="st">'bray'</span>, <span class="at">normalization =</span> <span class="st">'Log'</span>, <span class="at">filter_feature =</span> <span class="cn">FALSE</span>, <span class="at">feature_list =</span> <span class="cn">NULL</span>)</span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>jaccard <span class="ot">&lt;-</span> <span class="fu">GetBetaDiversity</span>(mmo, <span class="at">method =</span> <span class="st">'jaccard'</span>, <span class="at">normalization =</span> <span class="st">'Log'</span>, <span class="at">filter_feature =</span> <span class="cn">FALSE</span>, <span class="at">feature_list =</span> <span class="cn">NULL</span>)</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a><span class="co"># For weighted beta diversity, Generalized UniFrac can be used</span></span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a>guni <span class="ot">&lt;-</span> <span class="fu">GetBetaDiversity</span>(mmo, <span class="at">method =</span> <span class="st">'Gen.Uni'</span>, <span class="at">normalization =</span> <span class="st">'Log'</span>, <span class="at">distance =</span> <span class="st">'dreams'</span>, <span class="at">filter_feature =</span> <span class="cn">FALSE</span>, <span class="at">feature_list =</span> <span class="cn">NULL</span>)</span>
<span id="cb22-9"><a href="#cb22-9" tabindex="-1"></a>guni<span class="fl">.0</span> <span class="ot">&lt;-</span> guni[,,<span class="st">'d_0'</span>] <span class="co"># GUniFrac with alpha 0</span></span>
<span id="cb22-10"><a href="#cb22-10" tabindex="-1"></a>guni<span class="fl">.05</span> <span class="ot">&lt;-</span> guni[,,<span class="st">'d_0.5'</span>] <span class="co"># GUniFrac with alpha 0.5</span></span>
<span id="cb22-11"><a href="#cb22-11" tabindex="-1"></a>guni<span class="fl">.1</span> <span class="ot">&lt;-</span> guni[,,<span class="st">'d_1'</span>] <span class="co"># GUniFrac with alpha 1</span></span>
<span id="cb22-12"><a href="#cb22-12" tabindex="-1"></a></span>
<span id="cb22-13"><a href="#cb22-13" tabindex="-1"></a></span>
<span id="cb22-14"><a href="#cb22-14" tabindex="-1"></a><span class="co"># 9.2.1. NMDS plots for beta diversity</span></span>
<span id="cb22-15"><a href="#cb22-15" tabindex="-1"></a>nmds <span class="ot">&lt;-</span> <span class="fu">metaMDS</span>(guni<span class="fl">.05</span>, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">try =</span> <span class="dv">50</span>, <span class="at">trymax =</span> <span class="dv">100</span>)</span>
<span id="cb22-16"><a href="#cb22-16" tabindex="-1"></a>nmds_coords <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">scores</span>(nmds))</span>
<span id="cb22-17"><a href="#cb22-17" tabindex="-1"></a>groups <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb22-18"><a href="#cb22-18" tabindex="-1"></a><span class="cf">for</span> (col <span class="cf">in</span> <span class="fu">colnames</span>(mmo<span class="sc">$</span>feature_data)[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>)]) {</span>
<span id="cb22-19"><a href="#cb22-19" tabindex="-1"></a>  groups <span class="ot">&lt;-</span> <span class="fu">append</span>(groups, metadata[metadata<span class="sc">$</span>sample <span class="sc">==</span> col, ]<span class="sc">$</span>group)</span>
<span id="cb22-20"><a href="#cb22-20" tabindex="-1"></a>}</span>
<span id="cb22-21"><a href="#cb22-21" tabindex="-1"></a>nmds_coords<span class="sc">$</span>group <span class="ot">&lt;-</span> groups</span>
<span id="cb22-22"><a href="#cb22-22" tabindex="-1"></a></span>
<span id="cb22-23"><a href="#cb22-23" tabindex="-1"></a><span class="fu">ggplot</span>(nmds_coords, <span class="fu">aes</span>(<span class="at">x =</span> NMDS1, <span class="at">y =</span> NMDS2, <span class="at">color =</span> group)) <span class="sc">+</span></span>
<span id="cb22-24"><a href="#cb22-24" tabindex="-1"></a>      <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb22-25"><a href="#cb22-25" tabindex="-1"></a>      <span class="co">#geom_text_repel(aes(label = group), size = 3) +</span></span>
<span id="cb22-26"><a href="#cb22-26" tabindex="-1"></a>      <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb22-27"><a href="#cb22-27" tabindex="-1"></a>      <span class="fu">stat_ellipse</span>(<span class="at">level =</span> <span class="fl">0.90</span>) <span class="sc">+</span></span>
<span id="cb22-28"><a href="#cb22-28" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"NMDS1"</span>, <span class="at">y =</span> <span class="st">"NMDS2"</span>) <span class="sc">+</span></span>
<span id="cb22-29"><a href="#cb22-29" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span>
<span id="cb22-30"><a href="#cb22-30" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">'plots/betadiv/NDMS_guni05.pdf'</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img width="608" height="607" alt="Screenshot 2025-09-01 at 9 19 41 AM" src="https://github.com/user-attachments/assets/d18e868d-f934-4355-9d13-7dcf14b5bba0"></p>
<div class="sourceCode" id="cb23"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co"># 9.2.2. Quantification of beta diversity </span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>group_distances <span class="ot">&lt;-</span> <span class="fu">CalculateGroupBetaDistance</span>(mmo, <span class="at">beta_div =</span> guni<span class="fl">.05</span>, <span class="at">reference_group =</span> <span class="st">'ctrl'</span>, <span class="at">groups =</span> <span class="fu">c</span>(<span class="st">'le1'</span>, <span class="st">'sl1'</span>))</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="fu">ggplot</span>(group_distances, <span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> distance)) <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>      <span class="fu">geom_boxplot</span>(<span class="at">outlier.shape =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>      <span class="fu">geom_beeswarm</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>      <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Group"</span>, <span class="at">y =</span> <span class="st">"Beta Diversity"</span>)</span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">'plots/betadiv/group_dist.pdf'</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img width="609" height="608" alt="Screenshot 2025-09-01 at 9 19 55 AM" src="https://github.com/user-attachments/assets/3dcac3ac-35cf-427e-b8e2-4b462742e9e9"></p>
</div>
<div class="section level2">
<h2 id="id_10-exporting-compounds-of-interest">10. Exporting compounds of interest<a class="anchor" aria-label="anchor" href="#id_10-exporting-compounds-of-interest"></a>
</h2>
<p><code>ExportFeaturesToCSV(mmo, feature_list = DAMs_up$ctrl_vs_sl1.up, normalization = 'None', output_dir = 'output/sl1_up_features.csv')</code></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/AGPL-3" class="external-link">AGPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ecomet</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Dale L. Forrister <br><small class="roles"> Author </small>   </li>
<li>Min-Soo Choi <br><small class="roles"> Author, maintainer </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Dale L. Forrister, Min-Soo Choi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
